### PR TITLE
[android][cellular] Rewrite cellular module to kotlin

### PR DESCRIPTION
**Pull request by M1ST4KE**
# Why

Modules written in Java are being rewritten to Kotlin

# How

Generated Kotlin code from java.
Clean up code.

# Test Plan

Checked in the bare-expo app

### CODE REVIEW

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  override fun getName(): String = "ExpoCellular"
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    moduleRegistryDelegate.onCreate(moduleRegistry)
+  }
+
+  override fun getConstants(): HashMap<String, Any?> {
+    val constants = HashMap<String, Any?>()
+    constants["allowsVoip"] = SipManager.isVoipSupported(mContext)
+    val telephonyManager = mContext.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
```
Comment (ide) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:30]: Android does not guarantee this value is non-null so we need to handle that case. There are some cases like Instant Apps where some system services are null.

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    return constants
+  }
+
+  @SuppressLint("MissingPermission")
+  @ExpoMethod
+  fun getCellularGenerationAsync(promise: Promise) {
+    try {
+      val mSystemService = mContext.getSystemService(Context.TELEPHONY_SERVICE)
+      val mTelephonyManager = if (mSystemService != null) {
+        mSystemService as TelephonyManager
```
Comment (ide) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:52]: I'm not sure why the Java code had the `m` prefix for these local variables (perhaps in an earlier version of the module's first PR they were member variables and were not cleaned up properly) but we can safely remove it here.
```suggestion
      val systemService = mContext.getSystemService(Context.TELEPHONY_SERVICE)
      val telephonyManager = if (systemService != null) {
        systemService as TelephonyManager
```

**Review by ide — APPROVED** [review]

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+class CellularModule(
+  private val mContext: Context,
+  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate()
+) : ExportedModule(mContext), RegistryLifecycleListener {
+  override fun getName(): String = "ExpoCellular"
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    moduleRegistryDelegate.onCreate(moduleRegistry)
+  }
+
```
Comment (lukmccall) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:25]:
```suggestion
```
❔

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import org.unimodules.core.ModuleRegistry
+import org.unimodules.core.ModuleRegistryDelegate
+import org.unimodules.core.Promise
+import org.unimodules.core.interfaces.ExpoMethod
+import org.unimodules.core.interfaces.RegistryLifecycleListener
+import java.util.*
+
+class CellularModule(
+  private val mContext: Context,
+  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate()
```
Comment (lukmccall) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:18]:
```suggestion
```
❔

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      constants["carrier"] = telephonyManager.simOperatorName
+      constants["mobileCountryCode"] = telephonyManager.simOperator.substring(0, 3)
+      constants["mobileNetworkCode"] = StringBuilder(telephonyManager.simOperator).delete(0, 3).toString()
+    } else {
+      constants["carrier"] = null
+      constants["mobileCountryCode"] = null
+      constants["mobileNetworkCode"] = null
+    }
+    return constants
+  }
```
Comment (lukmccall) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:48]:
```suggestion
  override fun getConstants(): HashMap<String, Any?> = HashMap().apply {
    val systemService = mContext.getSystemService(Context.TELEPHONY_SERVICE)
    val telephonyManager = systemService as? TelephonyManager
    
    put("allowsVoip", SipManager.isVoipSupported(mContext))
    put("isoCountryCode", telephonyManager?.simCountryIso)

    // check if sim state is ready
    if (telephonyManager != null && telephonyManager.simState == TelephonyManager.SIM_STATE_READY) {
      put("carrier",  telephonyManager.simOperatorName)
      put("mobileCountryCode", telephonyManager.simOperator.substring(0, 3))
      put("mobileNetworkCode",  StringBuilder(telephonyManager.simOperator).delete(0, 3).toString())
    } else {
      put("carrier", null)
      put("mobileCountryCode", null)
      put("mobileNetworkCode", null)
    }
  }
```
or something like this

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        TelephonyManager.NETWORK_TYPE_EVDO_B,
+        TelephonyManager.NETWORK_TYPE_EHRPD,
+        TelephonyManager.NETWORK_TYPE_HSPAP -> {
+          promise.resolve(CellularGeneration.CG_3G.value)
+        }
+        TelephonyManager.NETWORK_TYPE_LTE -> {
+          promise.resolve(CellularGeneration.CG_4G.value)
+        }
+        else -> promise.resolve(CellularGeneration.UNKNOWN.value)
+      }
```
Comment (lukmccall) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:89]: extract to a separate function

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularPackage.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package expo.modules.cellular
+
+import android.content.Context
+import org.unimodules.core.BasePackage
+import org.unimodules.core.ExportedModule
+
+class CellularPackage : BasePackage() {
+  override fun createExportedModules(context: Context) =
+    listOf(CellularModule(context) as ExportedModule)
```
Comment (lukmccall) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularPackage.kt:9]:
```suggestion
    listOf(CellularModule(context))
```

**Review by lukmccall — CHANGES_REQUESTED** [review]

**Review by lukmccall — APPROVED** [review]
LGMT 💪

**Pull request merged**